File: backend/routes/agents.py
from fastapi import APIRouter, HTTPException
from repositories.agents import agent_repo
from utils.response import ApiResponse

router = APIRouter(prefix="/agents", tags=["Agents"])


@router.get("", response_model=ApiResponse)
async def get_agents():
    """Get all agents"""
    agents = agent_repo.get_all()
    return ApiResponse.ok(data=agents)


@router.get("/{agent_id}", response_model=ApiResponse)
async def get_agent(agent_id: int):
    """Get an agent by ID"""
    agent = agent_repo.get_by_id(agent_id)
    if not agent:
        raise HTTPException(status_code=404, detail="Agent not found")
    return ApiResponse.ok(data=agent)


@router.post("", response_model=ApiResponse)
async def create_agent(agent_in: dict):
    """Create a new agent"""
    agent = agent_repo.create(agent_in)
    return ApiResponse.ok(
        data=agent, message="Agent created"
    )


@router.put("/{agent_id}", response_model=ApiResponse)
async def update_agent(
    agent_id: int,
    agent_in: dict,
):
    """Update an agent"""
    agent = agent_repo.get_by_id(agent_id)
    if not agent:
        raise HTTPException(status_code=404, detail="Agent not found")
    updated = agent_repo.update(agent_id, agent_in)
    return ApiResponse.ok(
        data=updated, message="Agent updated"
    )


@router.delete("/{agent_id}", response_model=ApiResponse)
async def delete_agent(agent_id: int):
    """Delete an agent"""
    agent = agent_repo.get_by_id(agent_id)
    if not agent:
        raise HTTPException(status_code=404, detail="Agent not found")
    agent_repo.delete(agent_id)
    return ApiResponse.ok(message="Agent deleted")

File: backend/src/routes/agents.js
/**
 * Agent Routes
 */
import { AgentRepository } from '../repositories/agentRepository.js';

const repo = new AgentRepository();

/**
 * Format success response
 */
function successResponse(data = null, message = 'Success') {
  return { success: true, message, data };
}

/**
 * Format error response
 */
function errorResponse(message, error = null) {
  return { success: false, message, error, data: null };
}

/**
 * Register agent routes
 */
async function agentRoutes(fastify) {
  // Get all agents
  fastify.get('/', async (request, reply) => {
    try {
      const agents = await repo.findAll();
      return successResponse(agents);
    } catch (error) {
      request.log.error(error);
      return errorResponse('Failed to get agents');
    }
  });

  // Get agent by ID
  fastify.get('/:id', async (request, reply) => {
    try {
      const agentId = parseInt(request.params.id, 10);
      const agent = await repo.findById(agentId);
      if (!agent) {
        reply.code(404);
        return errorResponse('Agent not found');
      }
      return successResponse(agent);
    } catch (error) {
      request.log.error(error);
      reply.code(500);
      return errorResponse('Failed to get agent');
    }
  });

  // Create agent
  fastify.post('/', async (request, reply) => {
    try {
      const agent = await repo.create(request.body);
      return successResponse(agent, 'Agent created');
    } catch (error) {
      request.log.error(error);
      reply.code(500);
      return errorResponse('Failed to create agent');
    }
  });

  // Update agent
  fastify.put('/:id', async (request, reply) => {
    try {
      const agentId = parseInt(request.params.id, 10);
      const updated = await repo.update(agentId, request.body);
      if (!updated) {
        reply.code(404);
        return errorResponse('Agent not found');
      }
      return successResponse(updated, 'Agent updated');
    } catch (error) {
      request.log.error(error);
      reply.code(500);
      return errorResponse('Failed to update agent');
    }
  });

  // Delete agent
  fastify.delete('/:id', async (request, reply) => {
    try {
      const agentId = parseInt(request.params.id, 10);
      const deleted = await repo.delete(agentId);
      if (!deleted) {
        reply.code(404);
        return errorResponse('Agent not found');
      }
      return successResponse(null, 'Agent deleted');
    } catch (error) {
      request.log.error(error);
      reply.code(500);
      return errorResponse('Failed to delete agent');
    }
  });
}

export default agentRoutes;
